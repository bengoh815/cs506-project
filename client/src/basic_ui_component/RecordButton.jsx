import React from "react";

const RecordButton = () => {
  const startRecording = () => {
    console.log("Recording...");
    // Implement recording functionality
  };

  return (
    <div id="record">
      <h2 id="record-heading">Record</h2>
      <input
        type="button"
        id="record-file-input"
        value="Start Recording"
        onClick={startRecording}
      />
    </div>
  );
};

export default RecordButton;
