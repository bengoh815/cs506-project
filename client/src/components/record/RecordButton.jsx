/******************************************************************************
 * Filename: RecordButton.jsx
 * Purpose:  A component that allows the user to record audio.
 * Author:   Victor Nguyen & Don Ma
 *
 * Description:
 * This file contains the RecordButton component that allows the user to record
 * audio. The component displays a button that the user can click to start and
 * stop recording audio. The recorded audio is displayed to the user once the
 * recording has stopped and can be played back and downloaded.
 *
 * Usage:
 * To use this component, import it into the desired file and render it.
 *
 ******************************************************************************/

import React, { useState, useRef } from "react";

const RecordButton = () => {
  // State to store the MediaRecorder object
  const [mediaRecorder, setMediaRecorder] = useState(null);

  // State to store the recorded audio
  const [isRecording, setIsRecording] = useState(false);

  // State to store the recorded audio
  const [recordedAudio, setRecordedAudio] = useState(null);

  // Add a useRef hook to keep a reference to the stream
  const streamRef = useRef(null);

  /**
   * Requests access to the user's microphone. If access is granted, the MediaRecorder object is created and the recording starts.
   */
  const startRecording = () => {
    if (!navigator.mediaDevices) {
      console.error("User media not available");
      return;
    }

    // Reset the recorded audio
    setRecordedAudio(null);

    navigator.mediaDevices
      .getUserMedia({ audio: true })
      .then((stream) => {
        streamRef.current = stream; // Assign the stream to the ref
        const recorder = new MediaRecorder(stream);
        setMediaRecorder(recorder);

        recorder.start();
        setIsRecording(true);

        recorder.ondataavailable = (e) => {
          // e.data contains the audio data
          // You can create a new Blob object and create an object URL for it
          const url = URL.createObjectURL(e.data);
          console.log(url);

          setRecordedAudio(e.data);
        };

        recorder.onerror = (e) => console.error(e.error);
      })
      .catch((error) => {
        console.error(error);

        // Alert the user if they denied microphone access and prompt them to give permission in order to record audio
        if (error.name === "NotAllowedError") {
          alert(
            "Microphone permission denied. You need to give permission to record audio",
          );
        }
      });
  };

  /**
   * Stops the recording.
   */
  const stopRecording = () => {
    if (mediaRecorder) {
      mediaRecorder.stop();
      setIsRecording(false);
      streamRef.current.getTracks().forEach((track) => track.stop());
    }
  };

  return (
    <div id="record" data-testid="record">
      <h2 id="record-heading" data-testid="record-heading">
        Record
      </h2>
      <div style={{ display: "inline" }}>
        <input
          type="button"
          id="record-file-input"
          data-testid="record-file-input"
          value={isRecording ? "Stop Recording" : "Start Recording"}
          onClick={isRecording ? stopRecording : startRecording}
        />
      </div>

      {recordedAudio && !isRecording && (
        <>
          <br />
          <br />
          <audio src={URL.createObjectURL(recordedAudio)} controls />
        </>
      )}
    </div>
  );
};

export default RecordButton;
