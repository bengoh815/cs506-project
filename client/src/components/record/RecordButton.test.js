/******************************************************************************
 * Filename: RecordButton.test.js
 * Purpose:  Tests the RecordButton component.
 * Author:   Victor Nguyen
 *
 * Description:
 * This file contains tests for the RecordButton component. The tests ensure that
 * the component and its elements render correctly. A snapshot tests is included
 * to catch unintended changes to the component's structure or behavior.
 *
 * Usage:
 * Run the tests using the command `npm test`.
 *
 * Note:
 * More tests to be added/implemented to test actual functionality of the component
 * (e.g. recording audio).
 *
 ******************************************************************************/

import { render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";
import RecordButton from "./RecordButton";

test("whole record button component renders", () => {
  // Arrange
  render(<RecordButton />);
  const recordButtonElement = screen.getByTestId("record");

  // Assert
  expect(recordButtonElement).toBeInTheDocument();
});

test("record heading component renders", () => {
  // Arrange
  render(<RecordButton />);
  const recordHeadingElement = screen.getByTestId("record-heading");

  // Assert
  expect(recordHeadingElement).toBeInTheDocument();
});

test("correct record heading content", () => {
  // Arrange
  render(<RecordButton />);
  const recordHeadingElement = screen.getByTestId("record-heading");

  // Assert
  expect(recordHeadingElement).toHaveTextContent("Record");
});

test("record button component renders", () => {
  // Arrange
  render(<RecordButton />);
  const recordButtonElement = screen.getByTestId("record-file-input");

  // Assert
  expect(recordButtonElement).toBeInTheDocument();
});

test("correct record button content - start recording", () => {
  // Arrange
  render(<RecordButton />);
  const recordButtonElement = screen.getByTestId("record-file-input");

  // Assert
  expect(recordButtonElement).toHaveValue("Start Recording");
});

test.skip("correct record button content - stop recording", () => {
  // TODO: Implement test to check if the button content changes to "Stop Recording" when clicked
});

// Snapshot test
test("matches snapshot", () => {
  // Arrange
  const tree = renderer.create(<RecordButton />).toJSON();

  // Assert
  expect(tree).toMatchSnapshot();
});
