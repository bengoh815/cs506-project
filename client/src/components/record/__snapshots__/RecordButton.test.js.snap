// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  data-testid="record"
  id="record"
>
  <h2
    data-testid="record-heading"
    id="record-heading"
  >
    Record
  </h2>
  <div
    style={
      Object {
        "display": "inline",
      }
    }
  >
    <input
      data-testid="record-file-input"
      id="record-file-input"
      onClick={[Function]}
      type="button"
      value="Start Recording"
    />
  </div>
</div>
`;
